File: plugins/md-to-skill/hooks/instinct-suggest_pretooluse.py
#!/usr/bin/env python3
"""
md-to-skill PreToolUse Hook - Instinct Suggestion

Proactively suggests actions based on auto-approved instincts when tool use
matches instinct triggers. Never blocks execution.

Hooks on: Write|Edit|Bash|Read
"""

import json
import sys
import os
import re
import fnmatch
from datetime import datetime

# Try to import shared utilities from hook_utils (created by Package 1)
try:
    from hook_utils import parse_frontmatter, setup_plugin_path, load_hook_input
    HOOK_UTILS_AVAILABLE = True
except ImportError:
    HOOK_UTILS_AVAILABLE = False

# Fallback: setup plugin path manually if hook_utils not available
if HOOK_UTILS_AVAILABLE:
    setup_plugin_path()
else:
    PLUGIN_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    if PLUGIN_ROOT not in sys.path:
        sys.path.insert(0, PLUGIN_ROOT)

try:
    from config.config_loader import load_config, get_instinct_config
    CONFIG_AVAILABLE = True
except ImportError:
    CONFIG_AVAILABLE = False


def parse_instinct_frontmatter(content: str) -> dict:
    """Parse YAML frontmatter from an instinct markdown file.
    Fallback when hook_utils is not available."""
    if HOOK_UTILS_AVAILABLE:
        return parse_frontmatter(content)

    match = re.search(r'^---\s*\n(.*?)\n---', content, re.DOTALL | re.MULTILINE)
    if not match:
        return {}

    frontmatter_text = match.group(1)
    result = {}

    for line in frontmatter_text.split('\n'):
        line = line.strip()
        if not line or ':' not in line:
            continue

        key, _, value = line.partition(':')
        key = key.strip()
        value = value.strip()

        # Strip quotes
        if value.startswith('"') and value.endswith('"'):
            value = value[1:-1]
        elif value.startswith("'") and value.endswith("'"):
            value = value[1:-1]

        # Type conversions
        if value.lower() == 'true':
            result[key] = True
        elif value.lower() == 'false':
            result[key] = False
        elif re.match(r'^\d+$', value):
            result[key] = int(value)
        elif re.match(r'^\d+\.\d+$', value):
            result[key] = float(value)
        else:
            result[key] = value

    return result


def extract_action(content: str) -> str:
    """Extract the Action section from an instinct file."""
    match = re.search(r'## Action\s*\n(.+?)(?:\n##|\Z)', content, re.DOTALL)
    if match:
        return match.group(1).strip()
    return ''


def parse_match_patterns(content: str) -> list:
    """Extract match_patterns from instinct frontmatter content.

    match_patterns is a YAML list of dicts, e.g.:
    match_patterns:
      - tool: Write
        file_glob: "*.ts"
        path_contains: src/
      - tool: Bash
        command_prefix: npm
    """
    match = re.search(r'^---\s*\n(.*?)\n---', content, re.DOTALL | re.MULTILINE)
    if not match:
        return []

    fm_text = match.group(1)
    patterns = []
    current_pattern = None
    in_match_patterns = False

    for line in fm_text.split('\n'):
        stripped = line.strip()

        if stripped.startswith('match_patterns:'):
            in_match_patterns = True
            continue

        if in_match_patterns:
            # Check if we've left the match_patterns block (new top-level key)
            if not line.startswith(' ') and not line.startswith('\t') and ':' in stripped and not stripped.startswith('-'):
                break

            if stripped.startswith('- '):
                # New pattern entry
                if current_pattern:
                    patterns.append(current_pattern)
                current_pattern = {}
                # Parse inline key-value from "- tool: Write"
                rest = stripped[2:].strip()
                if ':' in rest:
                    k, _, v = rest.partition(':')
                    current_pattern[k.strip()] = v.strip().strip('"').strip("'")
            elif ':' in stripped and current_pattern is not None:
                # Continuation key-value
                k, _, v = stripped.partition(':')
                current_pattern[k.strip()] = v.strip().strip('"').strip("'")

    if current_pattern:
        patterns.append(current_pattern)

    return patterns


def load_auto_approved_instincts(cwd: str) -> list:
    """Load all auto-approved instincts from the instincts directory.
    Skips rejected instincts (M3)."""
    instincts_dir = os.path.join(cwd, '.claude', 'md-to-skill-instincts')
    instincts = []

    if not os.path.isdir(instincts_dir):
        return instincts

    try:
        for filename in os.listdir(instincts_dir):
            if not filename.endswith('.md'):
                continue

            filepath = os.path.join(instincts_dir, filename)
            try:
                with open(filepath, 'r', encoding='utf-8') as f:
                    content = f.read()

                fm = parse_instinct_frontmatter(content)

                # M3: Skip rejected instincts
                if fm.get('rejected', False):
                    continue

                if not fm.get('auto_approved', False):
                    continue

                action = extract_action(content)

                # M4: Extract match_patterns if present
                match_patterns = parse_match_patterns(content)

                instincts.append({
                    'id': fm.get('id', filename.replace('.md', '')),
                    'trigger': fm.get('trigger', ''),
                    'domain': fm.get('domain', ''),
                    'confidence': fm.get('confidence', 0),
                    'action': action,
                    'filepath': filepath,
                    'suggestions_shown': fm.get('suggestions_shown', 0),
                    'match_patterns': match_patterns,
                })
            except Exception:
                continue
    except Exception:
        pass

    return instincts


def match_by_patterns(match_patterns: list, tool_name: str, tool_input: dict) -> bool:
    """M4: Check if tool use matches explicit match_patterns.

    match_patterns is a list of dicts:
      - {tool: "Write|Edit", file_glob: "*.ts", path_contains: "src/"}
      - {tool: "Bash", command_prefix: "npm"}
      - {tool: "Read", file_glob: "*.config.*"}

    Returns True if any pattern matches.
    """
    for pattern in match_patterns:
        pattern_tool = pattern.get('tool', '')

        # Check if tool name matches (supports pipe-separated values)
        tool_options = [t.strip() for t in pattern_tool.split('|')]
        if tool_name not in tool_options:
            continue

        # For Write/Edit/Read — check file path
        if tool_name in ('Write', 'Edit', 'Read'):
            file_path = tool_input.get('file_path', '').replace('\\', '/')
            if not file_path:
                continue

            file_glob = pattern.get('file_glob', '')
            path_contains = pattern.get('path_contains', '')

            glob_match = True
            if file_glob:
                basename = os.path.basename(file_path)
                glob_match = fnmatch.fnmatch(basename, file_glob) or fnmatch.fnmatch(file_path, file_glob)

            path_match = True
            if path_contains:
                path_match = path_contains.lower() in file_path.lower()

            if glob_match and path_match:
                return True

        # For Bash — check command prefix
        elif tool_name == 'Bash':
            command = tool_input.get('command', '')
            command_prefix = pattern.get('command_prefix', '')

            if command_prefix and command.strip().lower().startswith(command_prefix.lower()):
                return True
            elif not command_prefix:
                # Pattern matched tool but no further constraint
                return True

    return False


def match_write_edit(instinct: dict, tool_input: dict) -> bool:
    """Check if an instinct trigger matches a Write/Edit tool context."""
    trigger = instinct.get('trigger', '').lower()
    file_path = tool_input.get('file_path', '').replace('\\', '/')
    if not file_path:
        return False

    file_lower = file_path.lower()
    basename = os.path.basename(file_lower)
    ext = os.path.splitext(basename)[1]

    # Match file extensions mentioned in trigger
    ext_patterns = re.findall(r'\.(ts|js|py|tsx|jsx|css|scss|html|json|md|cs|yaml|yml)', trigger)
    for ep in ext_patterns:
        if ext == '.' + ep:
            return True

    # Match path segments mentioned in trigger
    path_keywords = re.findall(r'(?:in|under|within|to)\s+(\S+)', trigger)
    for kw in path_keywords:
        if kw.strip('/').lower() in file_lower:
            return True

    # Match file type descriptions
    type_mappings = {
        'typescript': ['.ts', '.tsx'],
        'javascript': ['.js', '.jsx'],
        'python': ['.py'],
        'style': ['.css', '.scss', '.less'],
        'template': ['.html', '.hbs'],
        'test': ['.test.', '.spec.'],
        'component': ['.component.'],
        'service': ['.service.'],
        'module': ['.module.'],
    }
    for keyword, extensions in type_mappings.items():
        if keyword in trigger:
            for e in extensions:
                if e in file_lower:
                    return True

    # Match domain-based triggers
    domain = instinct.get('domain', '')
    if domain == 'naming' and ('writing' in trigger or 'creating' in trigger):
        return True
    if domain == 'code-style' and ('writing' in trigger or 'editing' in trigger):
        return True

    return False


def match_bash(instinct: dict, tool_input: dict) -> bool:
    """Check if an instinct trigger matches a Bash tool context."""
    trigger = instinct.get('trigger', '').lower()
    command = tool_input.get('command', '')
    if not command:
        return False

    cmd_lower = command.lower()
    first_token = cmd_lower.split()[0] if cmd_lower.strip() else ''

    # Match command names in trigger
    cmd_patterns = re.findall(r'(?:running|using|executing|after)\s+(\S+)', trigger)
    for cp in cmd_patterns:
        if cp.lower() == first_token:
            return True

    # Match common command keywords
    cmd_keywords = ['npm', 'git', 'python', 'node', 'bun', 'pnpm', 'yarn',
                    'dotnet', 'test', 'build', 'lint', 'deploy']
    for kw in cmd_keywords:
        if kw in trigger and kw in cmd_lower:
            return True

    # Match workflow triggers
    if 'test' in trigger and ('test' in cmd_lower or 'jest' in cmd_lower or 'vitest' in cmd_lower):
        return True
    if 'build' in trigger and 'build' in cmd_lower:
        return True

    return False


def match_read(instinct: dict, tool_input: dict) -> bool:
    """H4: Check if an instinct trigger matches a Read tool context.
    Only matches if match_patterns explicitly includes Read tool."""
    # Read matching only works via match_patterns (M4).
    # We don't do fuzzy trigger matching for Read to avoid noisy suggestions.
    return False


def increment_suggestions_shown(filepath: str, current_count: int):
    """Increment the suggestions_shown counter in instinct frontmatter."""
    try:
        with open(filepath, 'r', encoding='utf-8') as f:
            content = f.read()

        new_count = current_count + 1

        if 'suggestions_shown:' in content:
            content = re.sub(
                r'suggestions_shown:\s*\d+',
                f'suggestions_shown: {new_count}',
                content
            )
        else:
            # Add before the closing ---
            content = re.sub(
                r'\n---\n',
                f'\nsuggestions_shown: {new_count}\n---\n',
                content,
                count=1
            )

        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(content)
    except Exception:
        pass


def main():
    """Main entry point for the instinct suggestion hook."""
    try:
        # Load hook input
        if HOOK_UTILS_AVAILABLE:
            input_data = load_hook_input()
        else:
            input_data = json.load(sys.stdin)

        cwd = input_data.get('cwd', '.')
        tool_name = input_data.get('tool_name', '')
        tool_input = input_data.get('tool_input', {})

        if isinstance(tool_input, str):
            try:
                tool_input = json.loads(tool_input)
            except (json.JSONDecodeError, TypeError):
                tool_input = {}

        if not tool_name:
            print(json.dumps({"ok": True}))
            sys.exit(0)

        # Load auto-approved instincts (M3: rejected instincts filtered out)
        instincts = load_auto_approved_instincts(cwd)

        if not instincts:
            print(json.dumps({"ok": True}))
            sys.exit(0)

        # Find matching instincts
        matches = []
        for instinct in instincts:
            match_patterns = instinct.get('match_patterns', [])

            # M4: If match_patterns present, use them exclusively
            if match_patterns:
                matched = match_by_patterns(match_patterns, tool_name, tool_input)
            else:
                # Fall through to existing trigger-based matching (backward compat)
                matched = False
                if tool_name in ('Write', 'Edit'):
                    matched = match_write_edit(instinct, tool_input)
                elif tool_name == 'Bash':
                    matched = match_bash(instinct, tool_input)
                # H4: Read only matches via match_patterns, not trigger fallback

            if matched:
                matches.append(instinct)

        if not matches:
            print(json.dumps({"ok": True}))
            sys.exit(0)

        # Build suggestion message from matched instincts
        suggestions = []
        for m in matches:
            action = m.get('action', '')
            if action:
                suggestions.append(
                    f"[instinct:{m['id']}] ({m['domain']}, confidence:{m['confidence']}) {action}"
                )
                # Increment counter
                increment_suggestions_shown(m['filepath'], m.get('suggestions_shown', 0))

        if suggestions:
            message = "Instinct suggestions for this action:\n" + "\n".join(suggestions)
            print(json.dumps({"ok": True, "systemMessage": message}))
        else:
            print(json.dumps({"ok": True}))

        sys.exit(0)

    except Exception:
        # Never block on errors
        print(json.dumps({"ok": True}))
        sys.exit(0)


if __name__ == '__main__':
    main()

File: plugins/md-to-skill/hooks/instinct-suggest_pretooluse.js
#!/usr/bin/env node
/**
 * md-to-skill PreToolUse Hook - Instinct Suggestion
 *
 * Proactively suggests actions based on auto-approved instincts when tool use
 * matches instinct triggers. Never blocks execution.
 *
 * Hooks on: Write|Edit|Bash|Read
 */

const fs = require('fs');
const path = require('path');

const { parseFrontmatter, loadHookInputSync, findGitRoot, matchGlob } = require('./hook_utils');

let CONFIG_AVAILABLE = false;
let loadConfig, getInstinctConfig;
try {
  const pluginRoot = path.dirname(__dirname);
  const configLoader = require(pluginRoot + '/config/config_loader');
  loadConfig = configLoader.load_config || configLoader.loadConfig;
  getInstinctConfig = configLoader.get_instinct_config || configLoader.getInstinctConfig;
  CONFIG_AVAILABLE = true;
} catch { CONFIG_AVAILABLE = false; }


/**
 * Extract the Action section from an instinct file.
 * @param {string} content
 * @returns {string}
 */
function extractAction(content) {
  const match = content.match(/## Action\s*\n(.+?)(?:\n##|\s*$)/s);
  if (match) {
    return match[1].trim();
  }
  return '';
}


/**
 * Extract match_patterns from instinct frontmatter content.
 *
 * match_patterns is a YAML list of dicts, e.g.:
 * match_patterns:
 *   - tool: Write
 *     file_glob: "*.ts"
 *     path_contains: src/
 *   - tool: Bash
 *     command_prefix: npm
 *
 * @param {string} content
 * @returns {Array<object>}
 */
function parseMatchPatterns(content) {
  const match = content.match(/^---\s*\n([\s\S]*?)\n---/);
  if (!match) return [];

  const fmText = match[1];
  const patterns = [];
  let currentPattern = null;
  let inMatchPatterns = false;

  for (const line of fmText.split('\n')) {
    const stripped = line.trim();

    if (stripped.startsWith('match_patterns:')) {
      inMatchPatterns = true;
      continue;
    }

    if (inMatchPatterns) {
      // Check if we've left the match_patterns block (new top-level key)
      if (!line.startsWith(' ') && !line.startsWith('\t') && stripped.includes(':') && !stripped.startsWith('-')) {
        break;
      }

      if (stripped.startsWith('- ')) {
        // New pattern entry
        if (currentPattern) {
          patterns.push(currentPattern);
        }
        currentPattern = {};
        // Parse inline key-value from "- tool: Write"
        const rest = stripped.slice(2).trim();
        if (rest.includes(':')) {
          const colonIdx = rest.indexOf(':');
          const k = rest.substring(0, colonIdx).trim();
          const v = rest.substring(colonIdx + 1).trim().replace(/^["']|["']$/g, '');
          currentPattern[k] = v;
        }
      } else if (stripped.includes(':') && currentPattern !== null) {
        // Continuation key-value
        const colonIdx = stripped.indexOf(':');
        const k = stripped.substring(0, colonIdx).trim();
        const v = stripped.substring(colonIdx + 1).trim().replace(/^["']|["']$/g, '');
        currentPattern[k] = v;
      }
    }
  }

  if (currentPattern) {
    patterns.push(currentPattern);
  }

  return patterns;
}


/**
 * Load all auto-approved instincts from the instincts directory.
 * Skips rejected instincts (M3).
 * @param {string} cwd
 * @returns {Array<object>}
 */
function loadAutoApprovedInstincts(cwd) {
  const instinctsDir = path.join(cwd, '.claude', 'md-to-skill-instincts');
  const instincts = [];

  if (!fs.existsSync(instinctsDir)) {
    return instincts;
  }

  try {
    let stat;
    try { stat = fs.statSync(instinctsDir); } catch { return instincts; }
    if (!stat.isDirectory()) return instincts;

    const files = fs.readdirSync(instinctsDir);
    for (const filename of files) {
      if (!filename.endsWith('.md')) continue;

      const filepath = path.join(instinctsDir, filename);
      try {
        const content = fs.readFileSync(filepath, 'utf8');
        const fm = parseFrontmatter(content);

        // M3: Skip rejected instincts
        if (fm.rejected) continue;

        if (!fm.auto_approved) continue;

        const action = extractAction(content);

        // M4: Extract match_patterns if present
        const matchPatterns = parseMatchPatterns(content);

        instincts.push({
          id: fm.id || filename.replace('.md', ''),
          trigger: fm.trigger || '',
          domain: fm.domain || '',
          confidence: fm.confidence || 0,
          action: action,
          filepath: filepath,
          suggestions_shown: fm.suggestions_shown || 0,
          match_patterns: matchPatterns,
        });
      } catch {
        continue;
      }
    }
  } catch {
    // ignore
  }

  return instincts;
}


/**
 * M4: Check if tool use matches explicit match_patterns.
 *
 * match_patterns is a list of dicts:
 *   - {tool: "Write|Edit", file_glob: "*.ts", path_contains: "src/"}
 *   - {tool: "Bash", command_prefix: "npm"}
 *   - {tool: "Read", file_glob: "*.config.*"}
 *
 * Returns true if any pattern matches.
 * @param {Array<object>} matchPatterns
 * @param {string} toolName
 * @param {object} toolInput
 * @returns {boolean}
 */
function matchByPatterns(matchPatterns, toolName, toolInput) {
  for (const pattern of matchPatterns) {
    const patternTool = pattern.tool || '';

    // Check if tool name matches (supports pipe-separated values)
    const toolOptions = patternTool.split('|').map(t => t.trim());
    if (!toolOptions.includes(toolName)) continue;

    // For Write/Edit/Read — check file path
    if (['Write', 'Edit', 'Read'].includes(toolName)) {
      const filePath = (toolInput.file_path || '').replace(/\\/g, '/');
      if (!filePath) continue;

      const fileGlob = pattern.file_glob || '';
      const pathContains = pattern.path_contains || '';

      let globMatch = true;
      if (fileGlob) {
        const basename = path.basename(filePath);
        globMatch = matchGlob(basename, fileGlob) || matchGlob(filePath, fileGlob);
      }

      let pathMatch = true;
      if (pathContains) {
        pathMatch = filePath.toLowerCase().includes(pathContains.toLowerCase());
      }

      if (globMatch && pathMatch) return true;

    // For Bash — check command prefix
    } else if (toolName === 'Bash') {
      const command = toolInput.command || '';
      const commandPrefix = pattern.command_prefix || '';

      if (commandPrefix && command.trim().toLowerCase().startsWith(commandPrefix.toLowerCase())) {
        return true;
      } else if (!commandPrefix) {
        // Pattern matched tool but no further constraint
        return true;
      }
    }
  }

  return false;
}


/**
 * Check if an instinct trigger matches a Write/Edit tool context.
 * @param {object} instinct
 * @param {object} toolInput
 * @returns {boolean}
 */
function matchWriteEdit(instinct, toolInput) {
  const trigger = (instinct.trigger || '').toLowerCase();
  const filePath = (toolInput.file_path || '').replace(/\\/g, '/');
  if (!filePath) return false;

  const fileLower = filePath.toLowerCase();
  const basename = path.basename(fileLower);
  const ext = path.extname(basename);

  // Match file extensions mentioned in trigger
  const extPatterns = trigger.match(/\.(ts|js|py|tsx|jsx|css|scss|html|json|md|cs|yaml|yml)/g);
  if (extPatterns) {
    for (const ep of extPatterns) {
      if (ext === ep) return true;
    }
  }

  // Match path segments mentioned in trigger
  const pathKeywords = [...trigger.matchAll(/(?:in|under|within|to)\s+(\S+)/g)];
  for (const m of pathKeywords) {
    const kw = m[1];
    if (fileLower.includes(kw.replace(/\//g, '').toLowerCase())) {
      return true;
    }
  }

  // Match file type descriptions
  const typeMappings = {
    'typescript': ['.ts', '.tsx'],
    'javascript': ['.js', '.jsx'],
    'python': ['.py'],
    'style': ['.css', '.scss', '.less'],
    'template': ['.html', '.hbs'],
    'test': ['.test.', '.spec.'],
    'component': ['.component.'],
    'service': ['.service.'],
    'module': ['.module.'],
  };
  for (const [keyword, extensions] of Object.entries(typeMappings)) {
    if (trigger.includes(keyword)) {
      for (const e of extensions) {
        if (fileLower.includes(e)) return true;
      }
    }
  }

  // Match domain-based triggers
  const domain = instinct.domain || '';
  if (domain === 'naming' && (trigger.includes('writing') || trigger.includes('creating'))) {
    return true;
  }
  if (domain === 'code-style' && (trigger.includes('writing') || trigger.includes('editing'))) {
    return true;
  }

  return false;
}


/**
 * Check if an instinct trigger matches a Bash tool context.
 * @param {object} instinct
 * @param {object} toolInput
 * @returns {boolean}
 */
function matchBash(instinct, toolInput) {
  const trigger = (instinct.trigger || '').toLowerCase();
  const command = toolInput.command || '';
  if (!command) return false;

  const cmdLower = command.toLowerCase();
  const parts = cmdLower.trim().split(/\s+/);
  const firstToken = parts.length > 0 ? parts[0] : '';

  // Match command names in trigger
  const cmdPatterns = [...trigger.matchAll(/(?:running|using|executing|after)\s+(\S+)/g)];
  for (const m of cmdPatterns) {
    if (m[1].toLowerCase() === firstToken) return true;
  }

  // Match common command keywords
  const cmdKeywords = ['npm', 'git', 'python', 'node', 'bun', 'pnpm', 'yarn',
                       'dotnet', 'test', 'build', 'lint', 'deploy'];
  for (const kw of cmdKeywords) {
    if (trigger.includes(kw) && cmdLower.includes(kw)) return true;
  }

  // Match workflow triggers
  if (trigger.includes('test') && (cmdLower.includes('test') || cmdLower.includes('jest') || cmdLower.includes('vitest'))) {
    return true;
  }
  if (trigger.includes('build') && cmdLower.includes('build')) {
    return true;
  }

  return false;
}


/**
 * H4: Check if an instinct trigger matches a Read tool context.
 * Only matches if match_patterns explicitly includes Read tool.
 * @param {object} instinct
 * @param {object} toolInput
 * @returns {boolean}
 */
function matchRead(instinct, toolInput) {
  // Read matching only works via match_patterns (M4).
  // We don't do fuzzy trigger matching for Read to avoid noisy suggestions.
  return false;
}


/**
 * Increment the suggestions_shown counter in instinct frontmatter.
 * @param {string} filepath
 * @param {number} currentCount
 */
function incrementSuggestionsShown(filepath, currentCount) {
  try {
    let content = fs.readFileSync(filepath, 'utf8');
    const newCount = currentCount + 1;

    if (content.includes('suggestions_shown:')) {
      content = content.replace(
        /suggestions_shown:\s*\d+/,
        `suggestions_shown: ${newCount}`
      );
    } else {
      // Add before the closing ---
      content = content.replace(
        /\n---\n/,
        `\nsuggestions_shown: ${newCount}\n---\n`
      );
    }

    fs.writeFileSync(filepath, content, 'utf8');
  } catch {
    // ignore
  }
}


/**
 * Main entry point for the instinct suggestion hook.
 */
function main() {
  try {
    // Load hook input
    const inputData = loadHookInputSync();

    let cwd = inputData.cwd || '.';
    cwd = findGitRoot(cwd) || cwd;
    const toolName = inputData.tool_name || '';
    let toolInput = inputData.tool_input || {};

    if (typeof toolInput === 'string') {
      try {
        toolInput = JSON.parse(toolInput);
      } catch {
        toolInput = {};
      }
    }

    if (!toolName) {
      process.stdout.write(JSON.stringify({ ok: true }));
      process.exit(0);
    }

    // Load auto-approved instincts (M3: rejected instincts filtered out)
    const instincts = loadAutoApprovedInstincts(cwd);

    if (instincts.length === 0) {
      process.stdout.write(JSON.stringify({ ok: true }));
      process.exit(0);
    }

    // Find matching instincts
    const matches = [];
    for (const instinct of instincts) {
      const matchPatterns = instinct.match_patterns || [];

      let matched;
      // M4: If match_patterns present, use them exclusively
      if (matchPatterns.length > 0) {
        matched = matchByPatterns(matchPatterns, toolName, toolInput);
      } else {
        // Fall through to existing trigger-based matching (backward compat)
        matched = false;
        if (toolName === 'Write' || toolName === 'Edit') {
          matched = matchWriteEdit(instinct, toolInput);
        } else if (toolName === 'Bash') {
          matched = matchBash(instinct, toolInput);
        }
        // H4: Read only matches via match_patterns, not trigger fallback
      }

      if (matched) {
        matches.push(instinct);
      }
    }

    if (matches.length === 0) {
      process.stdout.write(JSON.stringify({ ok: true }));
      process.exit(0);
    }

    // Build suggestion message from matched instincts
    const suggestions = [];
    for (const m of matches) {
      const action = m.action || '';
      if (action) {
        suggestions.push(
          `[instinct:${m.id}] (${m.domain}, confidence:${m.confidence}) ${action}`
        );
        // Increment counter
        incrementSuggestionsShown(m.filepath, m.suggestions_shown || 0);
      }
    }

    if (suggestions.length > 0) {
      const message = 'Instinct suggestions for this action:\n' + suggestions.join('\n');
      process.stdout.write(JSON.stringify({ ok: true, systemMessage: message }));
    } else {
      process.stdout.write(JSON.stringify({ ok: true }));
    }

    process.exit(0);

  } catch {
    // Never block on errors
    process.stdout.write(JSON.stringify({ ok: true }));
    process.exit(0);
  }
}

main();
